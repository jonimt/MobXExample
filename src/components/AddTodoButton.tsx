import React, {FC} from 'react';
import {StyleSheet, ViewStyle, Pressable, Text} from 'react-native';

interface IAddTodoButton {
  onPress: () => void;
}

export const AddTodoButton: FC<IAddTodoButton> = ({onPress}) => {
  return (
    <Pressable testID={'AddTodoButton'} style={styles.button} onPress={onPress}>
      <Text>Add Todo</Text>
    </Pressable>
  );
};

const styles = StyleSheet.create({
  button: {
    backgroundColor: 'pink',
    height: 50,
    borderRadius: 25,
    marginTop: 8,
    marginHorizontal: 15,
    justifyContent: 'center',
    alignItems: 'center',
  } as ViewStyle,
});
