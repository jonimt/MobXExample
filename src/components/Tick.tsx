import React, {FC} from 'react';
import {View, StyleSheet, ViewStyle} from 'react-native';

interface ITick {
  completed: boolean;
}

export const Tick: FC<ITick> = ({completed}) => {
  const color = completed ? 'green' : 'red';

  return (
    <View
      testID={'TodoListItemTick'}
      style={[styles.tick, {backgroundColor: color}]}
    />
  );
};

const styles = StyleSheet.create({
  tick: {
    width: 20,
    height: 20,
    borderRadius: 8,
  } as ViewStyle,
});
