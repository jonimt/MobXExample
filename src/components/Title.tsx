import React, {FC} from 'react';
import {StyleSheet, Text, TextStyle} from 'react-native';

interface ITitle {
  title: string;
}

export const Title: FC<ITitle> = ({title}) => {
  return <Text style={styles.title}>{title}</Text>;
};

const styles = StyleSheet.create({
  title: {
    flex: 1,
    textAlign: 'center',
    fontFamily: 'HelveticaNeue-Medium',
    fontSize: 20,
  } as TextStyle,
});
