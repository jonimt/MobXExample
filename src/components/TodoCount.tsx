import React, {FC} from 'react';
import {StyleSheet, Text, TextStyle} from 'react-native';

interface ITodoCount {
  completed: number;
  total: number;
}

export const TodoCount: FC<ITodoCount> = ({completed, total}) => {
  return (
    <Text style={styles.todoCount}>
      {completed}/{total}
    </Text>
  );
};

const styles = StyleSheet.create({
  todoCount: {
    position: 'absolute',
    right: 15,
    fontFamily: 'HelveticaNeue',
    fontSize: 18,
  } as TextStyle,
});
