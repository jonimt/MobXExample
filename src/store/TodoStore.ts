import {action, computed, makeObservable, observable} from 'mobx';
import 'react-native-get-random-values';
import {v4 as uuid} from 'uuid';

export class Todo {
  id: string = uuid();
  title: string;
  completed: boolean = false;

  constructor(title: string) {
    this.title = title;
    makeObservable(this, {
      completed: observable,
      complete: action,
    });
  }

  complete() {
    this.completed = true;
  }
}

export class TodoStore {
  todos: Todo[] = [];

  constructor() {
    makeObservable(this, {
      todos: observable,
      totalTodos: computed,
      completedTodosCount: computed,
      addTodo: action,
    });
  }

  get totalTodos() {
    return this.todos.length;
  }

  get completedTodosCount() {
    return this.todos.filter(todo => todo.completed === true).length;
  }

  addTodo(title: string) {
    this.todos.push(new Todo(title));
  }
}

export const todoStore = new TodoStore();
