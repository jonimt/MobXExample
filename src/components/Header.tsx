import React, {FC} from 'react';
import {StyleSheet, View, ViewStyle} from 'react-native';
import {observer} from 'mobx-react-lite';
import {todoStore} from '../store/TodoStore';
import {TodoCount} from './TodoCount';
import {Title} from './Title';

export const Header: FC = observer(() => {
  return (
    <View style={styles.container}>
      <Title title={'Todos'} />
      <TodoCount
        completed={todoStore.completedTodosCount}
        total={todoStore.totalTodos}
      />
    </View>
  );
});

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: 50,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
    borderBottomWidth: 1,
    borderBottomColor: 'black',
  } as ViewStyle,
});
