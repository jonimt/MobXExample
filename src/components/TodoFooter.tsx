import React, {FC, useCallback, useState} from 'react';
import {StyleSheet, KeyboardAvoidingView, ViewStyle} from 'react-native';
import {todoStore} from '../store/TodoStore';
import {AddTodoButton} from './AddTodoButton';
import {TodoInput} from './TodoInput';

export const TodoFooter: FC = () => {
  const [todoTask, setTodoTask] = useState('');

  const onChangeText = useCallback((title: string) => setTodoTask(title), []);

  const addTodo = useCallback(() => {
    setTodoTask('');
    todoStore.addTodo(todoTask);
  }, [todoTask]);

  return (
    <KeyboardAvoidingView behavior={'padding'} style={styles.container}>
      <TodoInput value={todoTask} onChangeText={onChangeText} />
      <AddTodoButton onPress={addTodo} />
    </KeyboardAvoidingView>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
  } as ViewStyle,
});
