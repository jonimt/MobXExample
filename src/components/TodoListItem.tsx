import React, {FC} from 'react';
import {View, StyleSheet, ViewStyle, Pressable, Text} from 'react-native';
import {observer} from 'mobx-react-lite';
import {Todo} from '../store/TodoStore';
import {Tick} from './Tick';

interface ITodoListItem {
  todo: Todo;
}

export const TodoListItem: FC<ITodoListItem> = observer(({todo}) => {
  return (
    <View style={styles.container} testID={'TodoListItem'}>
      <Pressable
        testID={'CompleteTodoButton'}
        key={todo.id}
        style={styles.item}
        onPress={() => todo.complete()}>
        <View style={styles.flex1}>
          <Text numberOfLines={2}>{todo.title}</Text>
        </View>
        <Tick completed={todo.completed} />
      </Pressable>
    </View>
  );
});

const styles = StyleSheet.create({
  container: {
    width: '100%',
    paddingHorizontal: 15,
    alignItems: 'center',
  } as ViewStyle,
  flex1: {
    flex: 1,
  } as ViewStyle,
  item: {
    height: 50,
    marginTop: 8,
    paddingHorizontal: 8,
    borderWidth: 2,
    borderColor: 'brown',
    backgroundColor: 'pink',
    borderRadius: 16,
    justifyContent: 'space-between',
    alignItems: 'center',
    flexDirection: 'row',
  } as ViewStyle,
});
