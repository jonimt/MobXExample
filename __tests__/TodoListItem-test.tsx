import {fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import {TodoListItem} from '../src/components/TodoListItem';
import {Todo} from '../src/store/TodoStore';

// Creating a mock of the Todo
// and the complete method of the Todo
const mockComplete = jest.fn();
jest.mock('../src/store/TodoStore', () => {
  return {
    Todo: jest.fn().mockImplementation(() => {
      return {
        id: 'id',
        title: 'Test Todo',
        completed: false,
        complete: mockComplete,
      };
    }),
  };
});

// Clearing mock before every test
beforeEach(() => {
  Todo.mockClear();
});

describe('TodoListItem testing', () => {
  // Snapshot test
  it('renders correctly', () => {
    const tree = renderer
      .create(<TodoListItem todo={new Todo('Test Todo')} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  // Pressing the TodoListItem component should call
  // the complete method of the Todo passed as a prop
  it('Pressing TodoListItem is calling complete', () => {
    const {getByTestId} = render(<TodoListItem todo={new Todo('Test Todo')} />);

    // To make sure we target correct component, we use testID
    // check TodoListItem Pressable to see we have given it the id of CompleteTodoButton
    const pressable = getByTestId('CompleteTodoButton');
    fireEvent(pressable, 'press');

    expect(pressable).toBeTruthy();
    expect(mockComplete).toHaveBeenCalled();
  });
});
