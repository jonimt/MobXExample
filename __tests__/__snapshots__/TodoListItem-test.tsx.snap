// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoListItem testing renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "paddingHorizontal": 15,
      "width": "100%",
    }
  }
  testID="TodoListItem"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "pink",
        "borderColor": "brown",
        "borderRadius": 16,
        "borderWidth": 2,
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "marginTop": 8,
        "paddingHorizontal": 8,
      }
    }
    testID="CompleteTodoButton"
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        numberOfLines={2}
      >
        Test Todo
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "borderRadius": 8,
            "height": 20,
            "width": 20,
          },
          Object {
            "backgroundColor": "red",
          },
        ]
      }
      testID="TodoListItemTick"
    />
  </View>
</View>
`;
